include/deal.II/lac/petsc_ts.templates.h

```diff
@@ -401,12 +426,160 @@ namespace PETScWrappers
+          "prepare_for_coarsening_and_refinement",
+          __FILE__,
+          PETSC_ERR_LIB,
+          PETSC_ERROR_INITIAL,
+          "Failure in ts_resize_setup from dealii::PETScWrappers::TimeStepper");
+
+      PetscFunctionReturn(PETSC_SUCCESS);
+    };
+
+    const auto ts_resize_transfer = [](TS       ts,
+                                       PetscInt nv,
+                                       Vec      vin[],
+                                       Vec      vout[],
+                                       void *   ctx) -> PetscErrorCode {
+      PetscFunctionBeginUser;
+      auto user = static_cast<TimeStepper *>(ctx);
+
+      std::vector<VectorType> all_in, all_out;
+      for (PetscInt i = 0; i < nv; i++)
+        {
+          all_in.push_back(VectorType(vin[i]));
+        }
+
+      const int lineno = __LINE__;
+      const int err    = call_and_possibly_capture_ts_exception(
+        user->interpolate,
+        user->pending_exception,
+        []() -> void {},
+        all_in,
+        all_out);
+      if (err)
+        return PetscError(
+          PetscObjectComm((PetscObject)ts),
+          lineno + 1,
+          "interpolate",
+          __FILE__,
+          PETSC_ERR_LIB,
+          PETSC_ERROR_INITIAL,
+          "Failure in ts_resize_transfer from dealii::PETScWrappers::TimeStepper");
+
+      if (all_out.size() != all_in.size())
+        SETERRQ(PetscObjectComm((PetscObject)ts),
+                PETSC_ERR_LIB,
+                "Broken all_out");
+
+      for (PetscInt i = 0; i < nv; i++)
+        {
+          vout[i] = all_out[i].petsc_vector();
+          PetscCall(PetscObjectReference((PetscObject)vout[i]));
+        }
+
+      try
+        {
+          user->setup_callbacks();
+        }
+      catch (...)
+        {
+          SETERRQ(PetscObjectComm((PetscObject)ts),
+                  PETSC_ERR_LIB,
+                  "Exception in setup_callbacks");
+        }
+
+      try
+        {
+          user->setup_algebraic_constraints(all_out[0]);
+        }
+      catch (...)
+        {
+          SETERRQ(PetscObjectComm((PetscObject)ts),
+                  PETSC_ERR_LIB,
+                  "Exception in setup_algebraic_constraints");
+        }
+      PetscFunctionReturn(PETSC_SUCCESS);
+    };
+
+    const auto ts_poststep_amr = [](TS ts) -> PetscErrorCode {
+      PetscFunctionBeginUser;
       void *ctx;
+      PetscCall(TSGetApplicationContext(ts, &ctx));
+      auto user = static_cast<TimeStepper *>(ctx);
+
+      PetscErrorCode (
+        *transfer_setup)(TS, PetscInt, PetscReal, Vec, PetscBool *, void *);
+      PetscErrorCode (*transfer)(TS, PetscInt, Vec[], Vec[], void *);
+      AssertPETSc(PetscObjectQueryFunction((PetscObject)ts,
+                                           "__dealii_ts_resize_setup__",
+                                           &transfer_setup));
+      AssertPETSc(PetscObjectQueryFunction((PetscObject)ts,
+                                           "__dealii_ts_resize_transfer__",
```

Comment:
It took me quite a long time to figure out that this is really *declaring* two variables `transfer_setup` and `transfer`. Perhaps write this more easily as
```suggestion
      using transfer_setup_fn = 
      PetscErrorCode (*)(TS, PetscInt, PetscReal, Vec, PetscBool *, void *);
      transfer_setup_fn transfer;
```
and similarly for the other object.

Separately, why do you need to query PETSc for this? Could you not infer that by asking whether the callbacks of the current class are set to non-empty function objects?

Reply:
I have added the `using` clause